packages/web3-common/src/errors.ts

```diff
@@ -0,0 +1,333 @@
+		);
+		this.code = ERR_INVALID_RESPONSE;
+	}
+}
+
+export class TransactionError extends Web3Error {
+	public code = ERR_TX;
+
+	public constructor(message: string, public receipt: Receipt) {
+		super(message);
+	}
+
+	public toJSON() {
+		return { ...super.toJSON(), receipt: this.receipt };
+	}
+}
+
+export class RevertInstructionError extends Web3Error {
+	public code = ERR_TX_REVERT_INSTRUCTION;
+
+	public constructor(public reason: string, public signature: string) {
+		super(`Your request got reverted with the following reason string: ${reason}`);
+	}
+
+	public toJSON() {
+		return { ...super.toJSON(), reason: this.reason, signature: this.signature };
+	}
+}
+
+export class TransactionRevertError extends Web3Error {
+	public code = ERR_TX_REVERT_TRANSACTION;
+
+	public constructor(public reason: string, public signature: string, public receipt: Receipt) {
+		super(`Transaction has been reverted by the EVM:\n ${JSON.stringify(receipt, null, 2)}`);
+	}
+
+	public toJSON() {
+		return {
+			...super.toJSON(),
+			reason: this.reason,
+			signature: this.signature,
+			receipt: this.receipt,
+		};
+	}
+}
+
+export class NoContractAddressFoundError extends TransactionError {
+	public constructor(receipt: Receipt) {
+		super("The transaction receipt didn't contain a contract address.", receipt);
+		this.code = ERR_TX_NO_CONTRACT_ADDRESS;
+	}
+
+	public toJSON() {
+		return { ...super.toJSON(), receipt: this.receipt };
+	}
+}
+
+export class ContractCodeNotStoredError extends TransactionError {
+	public constructor(receipt: Receipt) {
+		super("The contract code couldn't be stored, please check your gas limit.", receipt);
+		this.code = ERR_TX_CONTRACT_NOT_STORED;
+	}
+}
+
+export class TransactionRevertedWithoutReasonError extends TransactionError {
+	public constructor(receipt: Receipt) {
+		super(
+			`Transaction has been reverted by the EVM:\n ${JSON.stringify(receipt, null, 2)}`,
+			receipt,
+		);
+		this.code = ERR_TX_REVERT_WITHOUT_REASON;
+	}
+}
+
+export class TransactionOutOfGasError extends TransactionError {
+	public constructor(receipt: Receipt) {
+		super(
+			`Transaction ran out of gas. Please provide more gas:\n ${JSON.stringify(
+				receipt,
+				null,
+				2,
+			)}`,
+			receipt,
+		);
+		this.code = ERR_TX_OUT_OF_GAS;
+	}
+}
+
+export class ResolverMethodMissingError extends Web3Error {
+	public code = ERR_CONTRACT_RESOLVER_MISSING;
+
+	public constructor(public address: string, public name: string) {
+		super(`The resolver at ${address} does not implement requested method: "${name}".`);
+	}
+
+	public toJSON() {
+		return { ...super.toJSON(), address: this.address, name: this.name };
+	}
+}
+
+export class ContractMissingABIError extends Web3Error {
+	public code = ERR_CONTRACT_ABI_MISSING;
+
+	public constructor() {
+		super(
+			'You must provide the json interface of the contract when instantiating a contract object.',
+		);
+	}
+}
+
+export class ContractOnceRequiresCallbackError extends Web3Error {
```

Comment:
We're not support callbacks?

Reply:
We have to provide callback support on `on` and `once` or `addEventListener` functions. But we should avoid using composite support fo `Promise` and `callback` e.g. in `call()` and `send()` functions. 

I am not sure this error would be used or not, I just moved all existing error constructs. If not used then we can remove it afterwards.